shared/controlflow/codeql/controlflow/Cfg.qll

```diff
@@ -0,0 +1,1293 @@
+
+  /** A split. */
+  class Split {
+    /** Gets a textual representation of this split. */
+    string toString();
+  }
+
+  /** A type of a control flow successor. */
+  class SuccessorType {
+    /** Gets a textual representation of this successor type. */
+    string toString();
+  }
+
+  /** Gets a successor type that matches completion `c`. */
+  SuccessorType getAMatchingSuccessorType(Completion c);
+
+  /**
+   * Hold if `t` represents simple (normal) evaluation of a statement or an
+   * expression.
+   */
+  predicate successorTypeIsSimple(SuccessorType t);
+
+  /** Hold if `t` represents a conditional successor type. */
+  predicate successorTypeIsCondition(SuccessorType t);
+
+  /** Holds if `t` is an abnormal exit type out of a CFG scope. */
+  predicate isAbnormalExitType(SuccessorType t);
+}
+
+/**
+ * Provides a shared interface for constructing control-flow graphs (CFGs) from
+ * abstract syntax trees (ASTs).
+ *
+ * The implementation is centered around the concept of a _completion_, which
+ * models how the execution of a statement or expression terminates.
+ *
+ * The CFG is built by structural recursion over the AST, using the abstract class
+ * `ControlFlowTree`. To achieve this, the CFG edges related to a given AST node,
+ * `n`, are divided into three categories:
+ *
+ *   1. The in-going edge that points to the first CFG node to execute when
+ *      `n` is going to be executed.
+ *   2. The out-going edges for control flow leaving `n` that are going to some
+ *      other node in the surrounding context of `n`.
+ *   3. The edges that have both of their end-points entirely within the AST
+ *      node and its children.
+ *
+ * The edges in (1) and (2) are inherently non-local and are therefore
+ * initially calculated as half-edges, that is, the single node, `k`, of the
+ * edge contained within `n`, by the predicates `k = n.first()` and `k = n.last(_)`,
+ * respectively. The edges in (3) can then be enumerated directly by the predicate
+ * `succ` by calling `first` and `last` recursively on the children of `n` and
+ * connecting the end-points. This yields the entire CFG, since all edges are in
+ * (3) for _some_ AST node.
+ *
+ * The parameter of `last` is the completion, which is necessary to distinguish
+ * the out-going edges from `n`. Note that the completion changes as the calculation of
+ * `last` proceeds outward through the AST; for example, a completion representing a
+ * loop break will be caught up by its surrounding loop and turned into a normal
+ * completion.
+ */
+module Make<LocationSig Location, InputSig<Location> Input> {
+  private import Input
+
+  final private class ControlFlowTreeBaseFinal = ControlFlowTreeBase;
+
+  /** An element with associated control flow. */
+  abstract class ControlFlowTree extends ControlFlowTreeBaseFinal {
+    /** Holds if `first` is the first element executed within this element. */
+    pragma[nomagic]
+    abstract predicate first(AstNode first);
```

Comment:
Would it be possible to replace most uses of `AstNode` with `ControlFlowTree(Base)` , something like:
```suggestion
    abstract predicate first(ControlFlowTreeBase first);
```